### ISSUE TITLE
Expose message class's canonical name

### ISSUE THREAD

**Issue by FND** [bufbuild/protobuf-es#739]
In order to do reliable message correlation (think request/response matching or dispatching via `oneof`), it would be helpful if message classes exposed a canonical name.

Given a message like this:

```javascript
import { proto3 } from "@bufbuild/protobuf";

const MyResponse = proto3.makeMessageType("my_package.MyResponse", () => [
    { no: 1, name: "request_id", kind: "scalar", T: 13 /* uint32 */ }
]);
```

... I'd like to convert `MyResponse` to `"myResponse"` - i.e. the same value used within `oneof` messages' `case` property. (I'm aware that such local names do not necessarily constitute unique identifiers, so one might argue that both full and local name was necessary.)

A bit of reverse engineering suggests I could use `localName({ kind: "rpc", name: MyResponse.name })` from a [private module](https://github.com/bufbuild/protobuf-es/blob/3be7c9d80750075e5bd25f1450e85a684811b627/packages/protobuf/src/private/names.ts#L33), but that seems ill-advised.

FWIW, I don't have strong feelings about instance/static property vs. a `(m: Message) => string` utility function.

PS: Thanks for your efforts on this project; it helped keep me comparatively sane.

**Comment by timostamm** [comment]
Hey @FND, the `localName` function is exported through the `codegenInfo` object:

```ts
import {codegenInfo} from "@bufbuild/protobuf";
codegenInfo.localName(...);
```

Note that it requires one of the [`Desc` types](https://github.com/bufbuild/protobuf-es/blob/main/docs/runtime_api.md#descriptors) as input.

The `codegenInfo` object also provides the functions `safeIdentifier` and `safeObjectProperty`. They are useful for escaping a Protobuf identifier for ECMAScript. 

We don't have a function for what you want specifically, but with the functions above, it should be implement a function that reliably converts message names any way you want.

**Comment by FND** [comment]
Thanks for the response! That sounds promising, but I have to admit it's not clear to me how I can get from a message class (`MyResponse` in my example) to such a descriptor. My own reading, exploration and reverse engineering got me nowhere; can you enlighten me?

**Comment by timostamm** [comment]
Descriptors are not available at runtime. They are available in plugins (see [docs](https://github.com/bufbuild/protobuf-es/blob/main/docs/writing_plugins.md)), but it looks like that's not what you're doing.

Messages have a type name. In your example, the type name is `"my_package.MyResponse"`. You can get it from a message instance via `message.getType().typeName`. You can strip the package prefix and lowercase the first character to get to `"myResponse"`.

I'm not sure that this is what you want. You mentioned the names in `oneof` fields. Note that these are all field names, not message names.

**Comment by FND** [comment]
You're right, I got myself confused there: Turns out I mostly just need the fields, not the message class itself. The puzzle pieces you've provided were helpful nonetheless; thank you!

Also, I was previously hesitant to derive identifiers from `typeName`, but I guess it shouldn't be an issue in practice.
